<a href="https://colab.research.google.com/github/Suchrahmadhani/DataScience-Practicum/blob/main/Pertemuan2_Suci_Rahmadhani_240401070508.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Notebook 2: Struktur Data Python, NumPy, dan Pandas

## Identitas Mahasiswa

| Keterangan | Informasi |
|------------|-----------|
| **Nama** | Suci Rahmadhani |
| **NIM** | 240401070508 |
| **Kelas** | IF404 |
| **Program Studi** | PJJ Informatika |
| **Mata Kuliah** | Data Science |

# Tujuan Praktikum

Pada praktikum ini saya mempelajari penggunaan struktur data Python serta library NumPy dan Pandas sebagai dasar dalam pengolahan data. Selain itu, saya juga mulai melakukan eksplorasi dataset Titanic untuk memahami cara membaca dataset, melihat informasi data, melakukan filtering, serta melakukan analisis sederhana menggunakan Pandas.

# Ringkasan Materi

Pada pertemuan ini saya belajar mengenai beberapa struktur data yang tersedia di Python, yaitu list, tuple, dictionary, dan set. Saya juga mulai mengenal NumPy sebagai library untuk komputasi numerik dan Pandas sebagai library yang digunakan untuk membaca, mengelola, serta menganalisis data dalam bentuk tabel (DataFrame).

Sebagai latihan, digunakan dataset Titanic untuk mempraktikkan berbagai operasi dasar seperti membaca data, mengecek informasi dataset, menghitung statistik sederhana, melakukan filtering, serta mengelompokkan data menggunakan fungsi-fungsi yang tersedia di Pandas.

In [1]:
import pandas as pd
import numpy as np

In [2]:
# ==========================================
# Python - List Operations
# ==========================================

# Membuat list
data = [10, 20, 'Python', True, 3.14]

# Akses & slicing
print(data[0])      # -> 10
print(data[-1])     # -> 3.14
print(data[1:3])    # -> [20, 'Python']

# Manipulasi
data.append(99)
data.insert(1, 55)
data.remove(20)

# data.sort()   # Tidak dijalankan karena list berisi tipe data campuran

print(data)

10
3.14
[20, 'Python']
[10, 55, 'Python', True, 3.14, 99]


In [3]:
# ==========================================
# Python - Tuple Operations
# ==========================================

koordinat = (10, 20, 30)

print(koordinat[0])
print(koordinat[-1])
print(len(koordinat))

10
30
3


In [4]:
# ==========================================
# Python - Dictionary Operations
# ==========================================

mahasiswa = {
    'nama': 'Suci Rahmadhani',
    'nim': '240401070508',
    'prodi': 'PJJ Informatika',
    'ipk': 3.98
}

print(mahasiswa['nama'])
print(mahasiswa.get('usia', 0))

mahasiswa['angkatan'] = 2024
del mahasiswa['nim']

print(mahasiswa)

Suci Rahmadhani
0
{'nama': 'Suci Rahmadhani', 'prodi': 'PJJ Informatika', 'ipk': 3.98, 'angkatan': 2024}


In [5]:
# ==========================================
# Python - Set Operations
# ==========================================

angka = {1, 2, 3, 3, 4, 5}

print(angka)

angka.add(6)

angka.remove(2)

print(angka)

{1, 2, 3, 4, 5}
{1, 3, 4, 5, 6}


In [6]:
# ==========================================
# Python - Membuat NumPy Array
# ==========================================

a1D = np.array([1,2,3,4,5])
a2D = np.array([[1,2,3],[4,5,6]])

zeros = np.zeros((3,4))
ones = np.ones((2,3))
eye = np.eye(4)
full = np.full((2,2),7)

arange = np.arange(0,20,2)
linsp = np.linspace(0,1,11)

rng = np.random.default_rng(seed=42)

rand_int = rng.integers(0,100,(3,4))
rand_flt = rng.normal(loc=0,scale=1,size=(100))

print(a1D)
print(a2D)

[1 2 3 4 5]
[[1 2 3]
 [4 5 6]]


In [7]:
# ==========================================
# NumPy Vectorisasi & Statistik
# ==========================================

a = np.array([1,2,3,4,5])
b = np.array([10,20,30,40,50])

print(a+b)
print(a*b)
print(a**2)

print(np.sqrt(a))
print(np.exp(a))
print(np.log(b))

data = rng.normal(0,1,1000)

print(f"Mean : {np.mean(data):.4f}")
print(f"Std  : {np.std(data):.4f}")
print(f"Min  : {np.min(data):.4f}")
print(f"Max  : {np.max(data):.4f}")
print(f"Q1   : {np.percentile(data,25):.4f}")
print(f"Q3   : {np.percentile(data,75):.4f}")

[11 22 33 44 55]
[ 10  40  90 160 250]
[ 1  4  9 16 25]
[1.         1.41421356 1.73205081 2.         2.23606798]
[  2.71828183   7.3890561   20.08553692  54.59815003 148.4131591 ]
[2.30258509 2.99573227 3.40119738 3.68887945 3.91202301]
Mean : -0.0163
Std  : 1.0008
Min  : -3.6484
Max  : 3.1789
Q1   : -0.6897
Q3   : 0.6163


In [8]:
# ==========================================
# Pandas Series
# ==========================================

s1 = pd.Series([85,92,78,95,88])

nilai = pd.Series(
    [85,92,78,95],
    index=['Alice','Bob','Caca','Dina']
)

print(nilai['Bob'])
print(nilai[nilai>80])
print(nilai.mean())
print(nilai.describe())

92
Alice    85
Bob      92
Dina     95
dtype: int64
87.5
count     4.000000
mean     87.500000
std       7.593857
min      78.000000
25%      83.250000
50%      88.500000
75%      92.750000
max      95.000000
dtype: float64


In [9]:
# ==========================================
# Pandas DataFrame
# ==========================================

data = {
    'nama':['Ali','Budi','Caca','Dina'],
    'usia':[22,25,21,23],
    'kota':['Jakarta','Bandung','Surabaya','Medan'],
    'ipk':[3.8,3.5,3.9,3.7]
}

df_mhs = pd.DataFrame(data)

print(df_mhs.shape)
print(df_mhs.dtypes)
print(df_mhs.columns)

(4, 4)
nama     object
usia      int64
kota     object
ipk     float64
dtype: object
Index(['nama', 'usia', 'kota', 'ipk'], dtype='object')


# Eksplorasi Dataset Titanic

In [10]:
url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"

df = pd.read_csv(url)

df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


In [11]:
df.info()

df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


# Operasi Dasar Pandas

In [12]:
# ==========================================
# Boolean Indexing
# ==========================================

penumpang_selamat = df[df['Survived'] == 1]

penumpang_wanita = df[df['Sex'] == 'female']

wanita_kelas1 = df[(df['Sex']=='female') & (df['Pclass']==1)]

df[['Name','Age','Sex','Survived']]

,Name,Age,Sex,Survived
0,"Braund, Mr. Owen Harris",22.0,male,0
1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",38.0,female,1
2,"Heikkinen, Miss. Laina",26.0,female,1
3,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",35.0,female,1
4,"Allen, Mr. William Henry",35.0,male,0
...,...,...,...,...
886,"Montvila, Rev. Juozas",27.0,male,0
887,"Graham, Miss. Margaret Edith",19.0,female,1
888,"Johnston, Miss. Catherine Helen ""Carrie""",NaN,female,0
889,"Behr, Mr. Karl Howell",26.0,male,1


In [13]:
# ==========================================
# GroupBy & Agregasi
# ==========================================

df.groupby('Pclass')['Survived'].mean()

hasil = df.groupby('Pclass').agg(
    total=('PassengerId','count'),
    selamat=('Survived','sum'),
    pct_selamat=('Survived','mean'),
    rata_usia=('Age','mean'),
    max_fare=('Fare','max')
)

hasil = hasil.round(3)

print(hasil)

df.groupby(['Pclass','Sex'])['Survived'] \
    .mean() \
    .unstack() \
    .round(3)

        total  selamat  pct_selamat  rata_usia  max_fare
Pclass                                                  
1         216      136        0.630     38.233   512.329
2         184       87        0.473     29.878    73.500
3         491      119        0.242     25.141    69.550


Sex,female,male
Pclass,,
1,0.968,0.369
2,0.921,0.157
3,0.500,0.135


# Hands-on Dataset Titanic (Q1–Q6)

Pada bagian ini saya menjawab seluruh pertanyaan latihan berdasarkan hasil eksplorasi dataset Titanic menggunakan library Pandas dan NumPy.

# Q1. Struktur Dataset Titanic

In [14]:
# Q1
# Menampilkan jumlah baris, kolom, dan daftar nama kolom

print(f"Jumlah baris dan kolom: {df.shape}")

print("\nDaftar Kolom:")
print(df.columns.tolist())

Jumlah baris dan kolom: (891, 12)

Daftar Kolom:
['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked']


### Interpretasi

Berdasarkan hasil di atas, dataset Titanic terdiri dari **891 baris** dan **12 kolom**. Setiap kolom menyimpan informasi yang berbeda mengenai identitas penumpang, kelas, jenis kelamin, usia, tarif, hingga status keselamatan. Informasi tersebut akan digunakan sebagai dasar untuk analisis pada pertanyaan berikutnya.

# Q2. Missing Values pada Dataset

In [15]:
# Q2
# Menampilkan jumlah dan persentase missing values pada setiap kolom

missing = df.isnull().sum()

persentase = (missing / len(df) * 100).round(2)

missing_df = pd.DataFrame({
    'Jumlah Missing': missing,
    'Persentase (%)': persentase
})

missing_df[missing_df['Jumlah Missing'] > 0]

,Jumlah Missing,Persentase (%)
Age,177,19.87
Cabin,687,77.10
Embarked,2,0.22


### Interpretasi

Hasil menunjukkan bahwa tidak semua kolom memiliki data yang lengkap. Kolom **Cabin** memiliki jumlah missing value paling banyak sehingga informasi kabin penumpang tidak tersedia untuk sebagian besar data. Selain itu, kolom **Age** juga memiliki cukup banyak data yang kosong, sedangkan kolom **Embarked** hanya memiliki sedikit missing value. Informasi ini penting karena data yang hilang perlu dipertimbangkan sebelum dilakukan analisis lebih lanjut.

# Q3. Persentase Penumpang yang Selamat

In [16]:
# Q3
# Menghitung persentase penumpang yang selamat

persentase_selamat = (df['Survived'].mean() * 100)

print(f"Persentase penumpang yang selamat: {persentase_selamat:.2f}%")

Persentase penumpang yang selamat: 38.38%


### Interpretasi

Berdasarkan hasil perhitungan, sekitar **38%** penumpang berhasil selamat dari kecelakaan Titanic. Artinya, jumlah penumpang yang tidak selamat masih jauh lebih banyak dibandingkan dengan yang selamat. Nilai ini memberikan gambaran awal mengenai kondisi keselamatan penumpang pada saat peristiwa tersebut terjadi.

# Q4. Penumpang Wanita Kelas 1

In [17]:
# Q4
# Menampilkan jumlah dan rata-rata usia penumpang wanita kelas 1

wanita_kelas1 = df[(df['Sex'] == 'female') & (df['Pclass'] == 1)]

print("Jumlah penumpang wanita kelas 1 :", len(wanita_kelas1))
print("Rata-rata usia :", round(wanita_kelas1['Age'].mean(), 2), "tahun")

Jumlah penumpang wanita kelas 1 : 94
Rata-rata usia : 34.61 tahun


### Interpretasi

Dari hasil filtering diketahui bahwa terdapat **94** penumpang wanita yang berada di kelas 1. Rata-rata usia mereka sekitar **34,61 tahun**. Hasil ini menunjukkan bagaimana Pandas dapat digunakan untuk memilih data berdasarkan lebih dari satu kondisi sekaligus.

# Q5. Tingkat Keselamatan Berdasarkan Kelas Penumpang

In [18]:
# Q5
# Menghitung tingkat keselamatan berdasarkan kelas penumpang

survival_rate = df.groupby('Pclass')['Survived'].mean()

print(survival_rate)

kelas_tertinggi = survival_rate.idxmax()

print(f"\nKelas dengan tingkat keselamatan tertinggi adalah Kelas {kelas_tertinggi}.")

Pclass
1    0.629630
2    0.472826
3    0.242363
Name: Survived, dtype: float64

Kelas dengan tingkat keselamatan tertinggi adalah Kelas 1.


### Interpretasi

Hasil analisis menunjukkan bahwa penumpang **kelas 1** memiliki tingkat keselamatan paling tinggi dibandingkan kelas lainnya. Sementara itu, tingkat keselamatan penumpang kelas 3 merupakan yang paling rendah. Hal ini menunjukkan bahwa kelas penumpang kemungkinan memiliki hubungan dengan peluang untuk selamat.

# Q6. Analisis Tambahan

In [19]:
# Q6
# Analisis tingkat keselamatan berdasarkan jenis kelamin

survival_gender = df.groupby('Sex')['Survived'].mean().round(3)

print(survival_gender)

Sex
female    0.742
male      0.189
Name: Survived, dtype: float64


### Interpretasi

Berdasarkan hasil analisis, tingkat keselamatan penumpang wanita jauh lebih tinggi dibandingkan penumpang pria. Hal ini menunjukkan bahwa jenis kelamin menjadi salah satu faktor yang berpengaruh terhadap peluang keselamatan penumpang Titanic. Dari hasil ini saya dapat melihat bagaimana fungsi **groupby()** dapat digunakan untuk membandingkan suatu variabel dengan lebih mudah.

# Kesimpulan

Pada praktikum ini saya mempelajari penggunaan library **NumPy** dan **Pandas** untuk melakukan pengolahan serta analisis data. Saya juga belajar membaca dataset dari file CSV, mengeksplorasi struktur data, melakukan filtering, menghitung statistik sederhana, serta menggunakan fungsi **groupby()** untuk memperoleh informasi dari dataset.

Melalui latihan menggunakan dataset Titanic, saya memahami bahwa proses analisis data tidak hanya berfokus pada melihat isi data, tetapi juga menemukan informasi yang dapat membantu dalam pengambilan kesimpulan. Saya masih perlu lebih banyak berlatih agar semakin terbiasa menggunakan berbagai fungsi di Pandas, terutama ketika menganalisis dataset yang lebih besar dan kompleks.